File: android/run/testSrc/com/android/tools/idea/run/tasks/RunInstantAppTest.java
/*
 * Copyright (C) 2018 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.android.tools.idea.run.tasks;

import static com.google.common.truth.Truth.assertThat;
import static org.mockito.Mockito.eq;
import static org.mockito.Mockito.isNull;
import static org.mockito.Mockito.any;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyNoMoreInteractions;
import static org.mockito.Mockito.when;

import com.android.ddmlib.IDevice;
import com.android.tools.idea.instantapp.InstantAppSdks;
import com.android.tools.idea.run.ApkFileUnit;
import com.android.tools.idea.run.ApkInfo;
import com.android.tools.idea.testing.AndroidProjectRule;
import com.android.tools.idea.testing.IdeComponents;
import com.google.android.instantapps.sdk.api.ExtendedSdk;
import com.google.android.instantapps.sdk.api.RunHandler;
import com.google.android.instantapps.sdk.api.StatusCode;
import com.google.common.collect.ImmutableList;
import com.intellij.execution.ExecutionException;
import com.intellij.execution.Executor;
import com.intellij.execution.process.ProcessHandler;
import com.intellij.execution.runners.ExecutionEnvironment;
import com.intellij.execution.ui.ConsoleView;
import com.intellij.openapi.progress.ProgressIndicator;
import com.intellij.openapi.project.Project;
import java.io.File;
import java.net.URL;
import java.util.Collections;
import junit.framework.TestCase;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;

public class RunInstantAppTest {
  private final String DEVICE_ID = "dev1234";
  private final File zipFile = new File("/tmp/fake.zip");
  private final ImmutableList<ApkInfo> apkInfoListForZip = ImmutableList.of(new ApkInfo(zipFile, "com.foo"));

  @Rule
  public AndroidProjectRule myProjectRule = AndroidProjectRule.inMemory();

  private InstantAppSdks instantAppSdks;
  @Mock private ExtendedSdk sdkLib;
  @Mock private RunHandler runHandler;
  private Project project;
  @Mock private Executor executor;
  @Mock private IDevice device;
  @Mock private ConsoleView consolePrinter;
  @Mock private ProcessHandler handler;
  @Mock private ProgressIndicator indicator;
  @Mock private ExecutionEnvironment env;

  @Before
  public void setUp() throws Exception {
    project = myProjectRule.getProject();
    MockitoAnnotations.initMocks(this);
    instantAppSdks = new IdeComponents(null, myProjectRule.getTestRootDisposable()).mockApplicationService(InstantAppSdks.class);
    when(instantAppSdks.loadLibrary()).thenReturn(sdkLib);
    when(sdkLib.getRunHandler()).thenReturn(runHandler);
    when(env.getProject()).thenReturn(project);
    when(env.getExecutor()).thenReturn(executor);
    when(device.getSerialNumber()).thenReturn(DEVICE_ID);
  }

  @Test
  public void testPerformWithNoZipFile() {
    RunInstantApp task = new RunInstantApp(ImmutableList.of(), "", Collections.singletonList(""));
    try {
      task.run(consolePrinter, device);
      TestCase.fail("Run should fail");
    }
    catch (ExecutionException e) {
      assertThat(e.getMessage()).isEqualTo("Uploading and launching Instant App: Package not found or not unique");
    }
    verifyNoMoreInteractions(runHandler);
  }

  @Test
  public void testPerformWithEmptyStringUrl() throws ExecutionException {
    RunInstantApp task = new RunInstantApp(apkInfoListForZip, "", Collections.singletonList(""));
    // Note here that an empty string URL should be transformed to null in the call to runInstantApp
    when(runHandler.runZip(
      /* zipFile= */ eq(zipFile),
      /* url= */ isNull(),
      /* adbHost= */ any(),
      /* adbDeviceId= */ eq(DEVICE_ID),
      /* runtimeApkDir= */ isNull(),
      /* resultStream= */ any(),
      /* progressIndicator= */ any()))
      .thenReturn(StatusCode.SUCCESS);
    task.run(consolePrinter, device);
  }

  @Test
  public void testPerformWithSpecifiedUrl() throws Exception {
    RunInstantApp task = new RunInstantApp(apkInfoListForZip, "http://foo.app", Collections.singletonList(""));
    when(runHandler.runZip(
      /* zipFile= */ eq(zipFile),
      /* url= */ eq(new URL("http://foo.app")),
      /* adbHost= */ any(),
      /* adbDeviceId= */ eq(DEVICE_ID),
      /* runtimeApkDir= */ isNull(),
      /* resultStream= */ any(),
      /* progressIndicator= */ any()))
      .thenReturn(StatusCode.SUCCESS);
    task.run(consolePrinter, device);
  }

  @Test
  public void testPerformWithListOfApks() throws Exception {
    File apk1 = new File("one.apk");
    File apk2 = new File("two.apk");
    File excludedApk = new File("excluded.apk");
    ImmutableList<ApkInfo> apkInfos = ImmutableList.of(
      new ApkInfo(ImmutableList.of(
        new ApkFileUnit("one", apk1),
        new ApkFileUnit("two", apk2),
        new ApkFileUnit("excluded", excludedApk)
      ), "com.dontcare"));

    RunInstantApp task = new RunInstantApp(apkInfos, "http://foo.app", ImmutableList.of("excluded"));

    when(runHandler.runApks(
      /* apkFiles= */ eq(ImmutableList.of(apk1, apk2)),
      /* url= */ eq(new URL("http://foo.app")),
      /* adbHost= */ any(),
      /* adbDeviceId= */ eq(DEVICE_ID),
      /* runtimeApkDir= */ isNull(),
      /* resultStream= */ any(),
      /* progressIndicator= */ any()))
      .thenReturn(StatusCode.SUCCESS);

    task.run(consolePrinter, device);

    verify(runHandler).runApks(
      /* apkFiles= */ eq(ImmutableList.of(apk1, apk2)),
      /* url= */ eq(new URL("http://foo.app")),
      /* adbHost= */ any(),
      /* adbDeviceId= */ eq(DEVICE_ID),
      /* runtimeApkDir= */ isNull(),
      /* resultStream= */ any(),
      /* progressIndicator= */ any());
  }
}

File: android/instantapp/testSrc/com/android/tools/idea/instantapp/RunInstantAppTest.kt
/*
 * Copyright (C) 2018 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.android.tools.idea.instantapp

import com.android.sdklib.AndroidVersion
import com.android.sdklib.devices.Abi
import com.android.testutils.MockitoCleanerRule
import com.android.tools.deployer.Deployer
import com.android.tools.idea.execution.common.ApplicationDeployer
import com.android.tools.idea.gradle.project.sync.snapshots.AndroidCoreTestProject
import com.android.tools.idea.projectsystem.applicationProjectContextForTests
import com.android.tools.idea.projectsystem.getProjectSystem
import com.android.tools.idea.run.AndroidRunConfiguration
import com.android.tools.idea.run.AndroidRunConfigurationExecutor
import com.android.tools.idea.run.FakeAndroidDevice
import com.android.tools.idea.run.configuration.execution.createApp
import com.android.tools.idea.testing.AndroidProjectRule
import com.android.tools.idea.testing.executeMakeBeforeRunStepInTest
import com.android.tools.idea.testing.gradleModule
import com.android.tools.idea.testing.mockDeviceFor
import com.google.android.instantapps.sdk.api.ExtendedSdk
import com.google.android.instantapps.sdk.api.RunHandler
import com.google.android.instantapps.sdk.api.StatusCode
import com.google.common.collect.ImmutableList
import com.google.common.truth.Truth.assertThat
import com.intellij.execution.RunManager
import com.intellij.execution.RunnerAndConfigurationSettings
import com.intellij.execution.configurations.ModuleBasedConfiguration
import com.intellij.execution.executors.DefaultRunExecutor
import com.intellij.execution.runners.ExecutionEnvironment
import com.intellij.execution.runners.ProgramRunner
import com.intellij.openapi.progress.EmptyProgressIndicator
import com.intellij.openapi.progress.ProgressManager
import com.intellij.openapi.util.Computable
import com.intellij.testFramework.replaceService
import java.io.File
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.mockito.kotlin.any
import org.mockito.kotlin.anyOrNull
import org.mockito.kotlin.argumentCaptor
import org.mockito.kotlin.eq
import org.mockito.kotlin.isNull
import org.mockito.kotlin.mock
import org.mockito.kotlin.never
import org.mockito.kotlin.times
import org.mockito.kotlin.verify
import org.mockito.kotlin.whenever

class RunInstantAppTest {
  private val device =
    mockDeviceFor(AndroidVersion(AndroidVersion.VersionCodes.O), listOf(Abi.X86_64, Abi.X86))
  private val applicationDeployer = mock<ApplicationDeployer>()

  private lateinit var runHandler: RunHandler
  private lateinit var configSettings: RunnerAndConfigurationSettings
  private val configuration
    get() = configSettings.configuration as AndroidRunConfiguration

  @get:Rule
  val projectRule =
    AndroidProjectRule.testProject(AndroidCoreTestProject.INSTANT_APP_WITH_DYNAMIC_FEATURES)

  @get:Rule val cleaner = MockitoCleanerRule()

  @Before
  fun setUp() {
    runHandler = mock<RunHandler>()

    val extendedSdk = mock<ExtendedSdk>()
    whenever(extendedSdk.runHandler).thenReturn(runHandler)

    val instantAppSdks =
      object : InstantAppSdks() {
        override fun loadLibrary(attemptUpgrades: Boolean): ExtendedSdk {
          return extendedSdk
        }
      }
    projectRule.replaceService(InstantAppSdks::class.java, instantAppSdks)
    configSettings =
      RunManager.getInstance(projectRule.project).allSettings.single {
        it.configuration is AndroidRunConfiguration
      }
  }

  private fun getExecutor(): AndroidRunConfigurationExecutor {
    val runner =
      ProgramRunner.getRunner(DefaultRunExecutor.EXECUTOR_ID, configSettings.configuration)!!
    val env =
      ExecutionEnvironment(
        DefaultRunExecutor.getRunExecutorInstance(),
        runner,
        configSettings,
        projectRule.project,
      )

    return AndroidRunConfigurationExecutor(
      (configSettings.configuration as ModuleBasedConfiguration<*, *>)
        .applicationProjectContextForTests,
      env,
      FakeAndroidDevice.forDevices(listOf(device)),
      projectRule.project.getProjectSystem().getApkProvider(configSettings.configuration)!!,
      applicationDeployer = applicationDeployer,
    )
  }

  @Test
  fun testDeployInstantAppAsInstantAPK() {

    // PREPARE
    configuration.DEPLOY_AS_INSTANT = true
    configuration.executeMakeBeforeRunStepInTest(device)

    whenever(
        runHandler.runApks(any(), anyOrNull(), anyOrNull(), anyOrNull(), anyOrNull(), any(), any())
      )
      .thenReturn(StatusCode.SUCCESS)

    // RUN
    ProgressManager.getInstance()
      .runProcess(
        Computable { getExecutor().run(ProgressManager.getInstance().progressIndicator) },
        EmptyProgressIndicator(),
      )

    val captor = argumentCaptor<ImmutableList<File>>()
    val serialNumber = device.serialNumber

    verify(runHandler, times(1))
      .runApks(captor.capture(), isNull(), anyOrNull(), eq(serialNumber), anyOrNull(), any(), any())

    val apks = (captor.firstValue as List<File>).map { it.name.substringAfterLast('/') }
    assertThat(apks)
      .containsExactly(
        "app-debug.apk",
        "instantdynamicfeature-debug.apk",
        "dynamicfeature-debug.apk",
      )
  }

  @Test
  fun testDeployInstantAppAsInstantAPKWithDisabledFeature() {

    // PREPARE
    configuration.DEPLOY_AS_INSTANT = true
    configuration.disabledDynamicFeatures =
      listOf(projectRule.project.gradleModule(":dynamicfeature")!!.name)
    configuration.executeMakeBeforeRunStepInTest(device)

    whenever(
        runHandler.runApks(any(), anyOrNull(), anyOrNull(), anyOrNull(), anyOrNull(), any(), any())
      )
      .thenReturn(StatusCode.SUCCESS)

    // RUN
    ProgressManager.getInstance()
      .runProcess(
        Computable { getExecutor().run(ProgressManager.getInstance().progressIndicator) },
        EmptyProgressIndicator(),
      )

    val captor = argumentCaptor<ImmutableList<File>>()
    val serialNumber = device.serialNumber

    verify(runHandler, times(1))
      .runApks(captor.capture(), isNull(), anyOrNull(), eq(serialNumber), anyOrNull(), any(), any())

    val apks = (captor.firstValue as List<File>).map { it.name.substringAfterLast('/') }
    assertThat(apks).containsExactly("app-debug.apk", "instantdynamicfeature-debug.apk")
  }

  @Test
  fun testDeployInstantAppAsInstantBundle() {

    // PREPARE
    configuration.DEPLOY_APK_FROM_BUNDLE = true
    configuration.DEPLOY_AS_INSTANT = true
    configuration.disabledDynamicFeatures =
      ImmutableList.of(projectRule.project.gradleModule(":dynamicfeature")!!.name)
    configuration.executeMakeBeforeRunStepInTest(device)

    whenever(
        runHandler.runApks(any(), anyOrNull(), anyOrNull(), anyOrNull(), anyOrNull(), any(), any())
      )
      .thenReturn(StatusCode.SUCCESS)

    // RUN
    ProgressManager.getInstance()
      .runProcess(
        Computable { getExecutor().run(ProgressManager.getInstance().progressIndicator) },
        EmptyProgressIndicator(),
      )

    val captor = argumentCaptor<ImmutableList<File>>()
    val serialNumber = device.serialNumber

    verify(runHandler, times(1))
      .runApks(captor.capture(), isNull(), anyOrNull(), eq(serialNumber), anyOrNull(), any(), any())

    val apks = (captor.firstValue as List<File>).map { it.name.substringAfterLast('/') }
    assertThat(apks)
      .containsExactly(
        "instant-base-master.apk",
        "instant-instantdynamicfeature-master.apk",
        "instant-base-mdpi.apk",
      )
  }

  @Test
  fun testDeployInstantAppWithoutInstantCheckbox() {
    configuration.executeMakeBeforeRunStepInTest(device)
    configuration.MODE = AndroidRunConfiguration.DO_NOTHING

    whenever(applicationDeployer.fullDeploy(eq(device), anyOrNull(), anyOrNull(), anyOrNull()))
      .thenReturn(
        Deployer.Result(
          false,
          false,
          false,
          createApp(device, "google.simpleapplication", emptyList(), emptyList()),
        )
      )

    // RUN
    ProgressManager.getInstance()
      .runProcess(
        Computable { getExecutor().run(ProgressManager.getInstance().progressIndicator) },
        EmptyProgressIndicator(),
      )

    verify(runHandler, never())
      .runZip(any(), anyOrNull(), anyOrNull(), anyOrNull(), anyOrNull(), any(), any())
    verify(runHandler, never())
      .runApks(any(), anyOrNull(), anyOrNull(), anyOrNull(), anyOrNull(), any(), any())
  }
}
